### PR TITLE
If opno of clause does not belong to opfamily of distributed key, do not use direct dispatch to resolve wrong result

### PR DESCRIPTION
**Pull request by zxuejing**
create table bar(a varchar);
insert into bar values('a   ');
explain select * from bar where a = 'a'::bpchar;
```
                                  QUERY PLAN
------------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.00 rows=1 width=8)
   ->  Seq Scan on bar  (cost=0.00..431.00 rows=1 width=8)
         Filter: ((a)::bpchar = 'a'::bpchar)
 Optimizer: Pivotal Optimizer (GPORCA)
(4 rows)

select * from bar where a = 'a'::bpchar;
 a
---
(0 rows)
```
we get wrong results when above sql is direct dispatched.

if opno of the clause does not belong to opfamily of distributed key,
do not use direct dispatch to avoid wrong results.

more details please refer to https://github.com/greenplum-db/gpdb/issues/14887

### CODE REVIEW

File: src/backend/optimizer/util/predtest_valueset.c [hunk 1]
```diff
@@ -380,6 +380,16 @@ TryProcessOpExprForPossibleValues(OpExpr *expr, Node *variable, PossibleValueSet
 		return false;
 	}
 
+	/*
+	 * if expr->opno does not belong to policy_opfamily of distributed key,
+	 * do not use direct dispatch to avoid wrong results.
+	 * more details please refer to https://github.com/greenplum-db/gpdb/issues/14887
+	 */
+	if (policy_opfamily != -1 && !op_in_opfamily(expr->opno, policy_opfamily))
```
Comment (kainwen) [inline src/backend/optimizer/util/predtest_valueset.c:388]: Recommend using `OidIsValid` here instread of `!= -1`.

Comment (zxuejing) [reply src/backend/optimizer/util/predtest_valueset.c:388]: thanks, i will modify.

File: src/backend/cdb/cdbtargeteddispatch.c [hunk 2]
```diff
@@ -172,7 +172,7 @@ GetContentIdsFromPlanForSingleRelation(PlannerInfo *root, Plan *plan, int rangeT
 							 GpSegmentIdAttributeNumber,
 							 vartypeid, type_mod, type_coll, 0);
 		pvs_segids = DeterminePossibleValueSet((Node *) qualification,
-											   (Node *) seg_id_var);
+											   (Node *) seg_id_var, InvalidOid);
```
Comment (kainwen) [inline src/backend/cdb/cdbtargeteddispatch.c:175]: Can we pass `interger_ops` family oid here?

File: src/backend/cdb/cdbtargeteddispatch.c [hunk 3]
```diff
@@ -171,8 +172,13 @@ GetContentIdsFromPlanForSingleRelation(PlannerInfo *root, Plan *plan, int rangeT
 		seg_id_var = makeVar(rangeTableIndex,
 							 GpSegmentIdAttributeNumber,
 							 vartypeid, type_mod, type_coll, 0);
+		Oid opclass = get_opclass_oid(HASH_AM_OID,
```
Comment (kainwen) [inline src/backend/cdb/cdbtargeteddispatch.c:175]: I think we can call this function:

```c
GetDefaultOpClass()
```

here.

**Review by kainwen — APPROVED** [review]

**Review by Aegeaner — APPROVED** [review]
LGTM

**Comment by kainwen** [comment]
@zxuejing Please hold on, let me think more on this.

**Comment by kainwen** [comment]
Spend much time on the PR's case change of `citext`.

```sql
create table t(a citext);
explain select * from t where a = 'a'::text; -- this will not direct dispatch with this PR's patch
```

I try to see if Postgres can do this kind of hash partition prune:

```sql
-- pg 16
create table t (a citext) partition by hash (a);
create table t0 partition of t for values with (modulus 3,remainder 0);
create table t1 partition of t for values with (modulus 3,remainder 1);
create table t2 partition of t for values with (modulus 3,remainder 2);

explain select * from t where a = 'a'::text;

                          QUERY PLAN
--------------------------------------------------------------
 Append  (cost=0.00..81.11 rows=21 width=32)
   ->  Seq Scan on t0 t_1  (cost=0.00..27.00 rows=7 width=32)
         Filter: ((a)::text = 'a'::text)
   ->  Seq Scan on t1 t_2  (cost=0.00..27.00 rows=7 width=32)
         Filter: ((a)::text = 'a'::text)
   ->  Seq Scan on t2 t_3  (cost=0.00..27.00 rows=7 width=32)
         Filter: ((a)::text = 'a'::text)
(7 rows)
```

We can see Latest Postgres also does not do partitune prune for this kind of QUAL.

Now come back to the GP's SQL, the qual is:

```
OpExpr [opno=98 opfuncid=67 opresulttype=16 opretset=false inputcollid=100]
        RelabelType [resulttype=25 resultcollid=100 relabelformat=COERCE_IMPLICIT_CAST]
                Var [varno=1 varattno=1 vartype=16386 varcollid=100 varnoold=1 varoattno=1]
        Const [consttype=25 constcollid=100 constlen=-1 constvalue=94126879949328 constisnull=false constbyval=false]
```

The `=` here is a `texteq`, does not belong to the hash opfamily of the table's distkey.

But, we, from the implementation can do the following deduction:

1. `texteq` ==> `citext_eq`
2. with `citext_eq` we can do the direct dispatch deduction

Seems we do not have the kind of implication rule in Postgres: `texteq` ==> `citext_eq`.

-------

In short, I am OK to push this PR with work:

1. add the above analysis of citext related case in commit message and in the test case change place
2. add the partition prune as a guard test case

@zxuejing Thanks for you contribution!
